File: back/app/models/transaction.rb
# frozen_string_literal: true

class Transaction < ApplicationRecord
  belongs_to :user

  # 取引タイプを定義（string列のためキー文字列で格納）
  enum :transaction_type, {
    income: 'income',    # 収入（貯金）
    expense: 'expense'   # 支出
  }

  # 指定した日を含む月の支出（予算の消化状況・実績の判定で使う）
  scope :expense_in_month, ->(date) { expense.where(date: date.all_month) }

  validates :amount, presence: true, numericality: { greater_than: 0 }
  validates :description, presence: true
  validates :transaction_type, presence: true
  validates :date, presence: true
end

File: front/src/types/transaction.ts
/** Rails API レスポンス型（スネークケース） */
export type ApiTransaction = {
  id: number
  amount: string | number
  transaction_type: string
  category: string
  description: string | null
  date: string
}

/** フロント内部型（キャメルケース） */
export type Transaction = {
  id: string
  amount: number
  type: "income" | "expense"
  category: string
  description: string
  date: string
}

/** ApiTransaction -> Transaction に変換する */
export const mapApiTransaction = (t: ApiTransaction): Transaction => ({
  id: String(t.id),
  amount: Number(t.amount),
  type: t.transaction_type as "income" | "expense",
  category: t.category,
  description: t.description || "",
  date: t.date ? t.date.split("T")[0] : "",
})
